### DISCUSSION TITLE
Is it possible to execute an external JS script with a single-file executable?

### DISCUSSION THREAD

**Discussion by mnapoli** [oven-sh/bun#12235] [Q&A]
Imagine I compile a CLI using Bun into a standalone executable (https://bun.sh/docs/bundler/executables).

Users install my binary by downloading it on their machine.

Then, I want my CLI to execute a JS file using the Bun runtime that is bundled in this executable.

E.g. instead of running:

```sh
bun run ./script.js
```

it would look like this:

```sh
my-cli-containing-bun
# this script runs "bun ./script.js" with "bun" being the engine from my CLI
```

Is that possible? If yes, how?

I would guess that I need to spawn a subprocess, but instead of spawning `bun run` I need to run the internal Bun engine…

**Comment by Bellisario** [comment]
You could try with a [dynamic import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) like so:

```js
// use user-passed file
const givenPath = process.argv[2];

// dynamic import
// warning: configured like so it only runs files in the current project folder
// you should probably consider converting relative paths to absolute when importing
const { default: importedFn } = await import(`./${givenPath}`);

importedFn();
```

Compile your code:

```bash
bun build ./index.ts --compile --outfile index
```

Run JS/TS files that export a default function:

```js
export default function() {
    console.log('Hello, world!');
}
```

```bash
$ ./index example.ts
Hello, world!
```

> [!WARNING]
> While this might be a solution, you should always remember you're not actually creating a different process but you're importing and executing a function directly on your main process, so make sure to **only run trusted code** in this way.

**Reply by mnapoli** [reply]
Ha 😄 yes thank you, that's exactly what I tried at first!

In theory it works, I can even sub-spawn the current process with the extra argument if I wanted to run in an isolated sub-process.

BUT the problem is the script I'm running actually behaved differently when importing/requiring it, than when running it with `bun run script.js`. I have no idea why, but Bun must run things slightly differently, so that doesn't work here.

**Reply by guest271314** [reply]
> BUT the problem is the script I'm running actually behaved differently when importing/requiring it

What is different?

**Comment by mwc** [comment]
@mnapoli 
maybe you want to run commands like `bun add` or `bun install`?

look like:
```bash
$ my-cli-containing-bun add -D dayjs
```

try bundling Bun together, but resulting file size will be larger,
first, you need to download the ZIP file corresponding to your platform from [this link](https://github.com/oven-sh/bun/releases) into your current working directory, then extract it (it contains a single bun file).
example: 
platform bun-linux-arm64 => download bun-linux-aarch64.zip

```bash
$ bun build --compile --target=bun-linux-arm64 ./index.ts ./bun-linux-aarch64/bun --outfile my-cli-containing-bun
```

entry file(index.ts):
```ts
import { $, argv, embeddedFiles, file, readableStreamToText, spawn, write } from 'bun'

// check if additional parameters were passed, for example, running ./my-cli-containing-bun a.ts
if (argv.length > 2) {
    // Write the asset file - the Bun CLI, into a directory within the current one, such as `.bin`
    // and give bun executable permissions.
    if (!await file('./.bin/bun').exists()) {
        await write(file('./.bin/bun'), embeddedFiles[0])
        await $`chmod 777 ./.bin/bun`
    }

    // as the Bun CLI to execute.
    const { stdout } = await spawn(['./.bin/bun', ...argv.slice(2)])
    console.log(await readableStreamToText(stdout))
}
else {
    // other cases, proceed with normal code execution.
    console.log('normal code...')
}
```

of course, this is not the optimal solution.
the `$` command embeds Bun, for example, 
```ts
await $`bun a.ts`
```
, but for a single executable file, this Bun is `my-cli-containing-bun`.

you can easily get the asset file via `file` function, or get the `entry file` through a virtual path `/$bunfs/root/.`(argv[1]). 

however, there is no documentation explaining how to get the embedded Bun CLI; otherwise, there would be no need to bundle the Bun CLI as an asset file.

**Comment by guest271314** [comment]
> Users install my binary by downloading it on their machine.

When testing I've found that the resulting executable still depends on the source files that the executable was bundled from to be on the machine, see https://github.com/oven-sh/bun/issues/14676.
